### PR TITLE
Add opcode `json ref`

### PR DESCRIPTION
**Pull request by thdailong**
Closes #x

## Proposed Changes

- Add opcode `json ref`

## TODO

- [x] update change log

### CODE REVIEW

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 1]
```diff
@@ -5087,3 +5091,112 @@ export class Replace3 extends Replace {
...
+	computeCost(): number {
+		return 25 + 2 * Math.ceil(this.length / 7); // cost = 25 + ceil(bytes / 7)
+	}
+	
+	execute(stack: TEALStack): number {
+		this.assertMinStackLen(stack, 2, this.line);
+		const key = this.assertBytes(stack.pop(), this.line);
+		const object = this.assertBytes(stack.pop(), this.line);
+		this.length = object.length;
+		const enc = new TextDecoder("utf-8");
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:5146]: maybe lets name it 'decoder' or 'utf8decoder'

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 2]
```diff
@@ -5087,3 +5091,112 @@ export class Replace3 extends Replace {
...
+		const decoded = enc.decode(object);
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:5147]: lets follow the same convention both for object and the key. Maybe 'decodedObj' and 'decodedKey'?

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 3]
```diff
@@ -7283,4 +7284,51 @@ describe("Teal Opcodes", function () {
 			assert.equal(1900, op.execute(stack));
 		});
 	});
+
+	describe("json_ref", function () {
+		const stack = new Stack<StackElem>();
+		const jsonByte = "{\"key0\": 0,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key4\": {\"key40\": 10}}, \"key5\": 18446744073709551615 }";
+
+		it("should return correct JSONUint64", function () {
```
Comment (sczembor) [inline packages/runtime/test/src/interpreter/opcode-list.ts:7292]:
```suggestion
		it("Should return correct JSONUint64", function () {
```
and lets change it in all the cases

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 4]
```diff
@@ -7283,4 +7284,51 @@ describe("Teal Opcodes", function () {
...
+			stack.push(parsing.stringToBytes("{\"maxUint64\": 18446744073709551615}"));
+			stack.push(parsing.stringToBytes("maxUint64"));
+			const op = new Json_ref(["JSONUint64"], 1);
+			op.execute(stack);
+
+			const top = stack.pop();
+			assert.deepEqual(18446744073709551615n, top);
+		});
+
+		it("should throw when get wrong type JSON(expect byte but got uint64)", function () {
```
Comment (sczembor) [inline packages/runtime/test/src/interpreter/opcode-list.ts:7302]:
```suggestion
		it("should throw when get wrong JSON type(expect byte but got uint64)", function () {
```

File: packages/runtime/src/lib/parsing.ts [hunk 5]
```diff
@@ -282,4 +283,19 @@ export function bigEndianBytesToBigInt(bytes: Uint8Array | Buffer): bigint {
  */
 export function strHexToBytes(str: string): Uint8Array {
 	return new Uint8Array(Buffer.from(str.slice(2), "hex"));
+}
+
+/**
+ * assert if string given is a valid JSON object
```
Comment (sczembor) [inline packages/runtime/src/lib/parsing.ts:289]:
```suggestion
 * assert if given string is a valid JSON object
```

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 2 above)
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:5147]: `decodedObj`

**Review by sczembor — COMMENTED** [review]
Looks good. Let m know if we cover all the important test cases from go-algorand

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 6]
```diff
@@ -7283,4 +7284,51 @@ describe("Teal Opcodes", function () {
 			assert.equal(1900, op.execute(stack));
 		});
 	});
+
+	describe("json_ref", function () {
```
Comment (sczembor) [inline packages/runtime/test/src/interpreter/opcode-list.ts:7288]:
```suggestion
	describe("json_ref", ()=> {
```

Comment (sczembor) [reply packages/runtime/test/src/interpreter/opcode-list.ts:7288]: maybe lets change this in all the cases. Its shorter

File: packages/runtime/src/lib/parsing.ts [hunk 7]
```diff
@@ -282,4 +283,19 @@ export function bigEndianBytesToBigInt(bytes: Uint8Array | Buffer): bigint {
...
+ * @param jsonString
+ */
+ export function assertJSON(jsonString: string, line: number): void {
+	const strictBigJSON = JSONbig({ strict: true });
+	try {
+		strictBigJSON.parse(jsonString);
+	} catch (e) {
```
Comment (sczembor) [inline packages/runtime/src/lib/parsing.ts:296]: do we catch all the error here?

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -5087,3 +5091,112 @@ export class Replace3 extends Replace {
...
+					throw new RuntimeError(RUNTIME_ERRORS.TEAL.INVALID_TYPE, {
+						expected: 'object',
+						actual: typeof jsonObject[keyString],
+						line: this.line,
+					});
+				} 
+				break;
+			}
+		}
+
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:5199]:
```suggestion
```

File: packages/runtime/src/lib/parsing.ts (hunk 7 above)
Comment (thdailong) [reply packages/runtime/src/lib/parsing.ts:296]: Yes we catch all the errors from JSON.parse

Comment (sczembor) [reply packages/runtime/src/lib/parsing.ts:296]: Should we handle them? To give a better feedback to the user when it throws an error

Comment (thdailong) [reply packages/runtime/src/lib/parsing.ts:296]: Well this one function `assertJSON` just to make sure the bytes we parse is correct JSON format. Incase the string was wrong, we throw error invalid json parsing

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 9]
```diff
@@ -7283,4 +7284,51 @@ describe("Teal Opcodes", function () {
...
+			op1.execute(stack);
+			stack.push(parsing.stringToBytes("key4"));
+			const op2 = new Json_ref(["JSONObject"], 1);
+			op2.execute(stack);
+
+			const top = stack.pop();
+			const expected = parsing.stringToBytes("{\"key40\":10}");
+			assert.deepEqual(top, expected);
+		});
+	});
```
Comment (robert-zaremba) [inline packages/runtime/test/src/interpreter/opcode-list.ts:7333]: please make sure you added few tests from go-algorand.

**Review by sczembor — APPROVED** [review]
LGTM

**Pull request merged**